sway-fmt-v2/src/utils/comments.rs

```diff
@@ -0,0 +1,169 @@
+use crate::fmt::FormatterError;
+use std::{cmp::Ordering, collections::BTreeMap, sync::Arc};
+use sway_parse::token::{lex_commented, Comment, CommentedTokenTree, CommentedTree};
+
+/// Represents a span for the comments in a spesific file
+/// A stripped down version of sway-types::src::Span
+#[derive(PartialEq, Eq, Debug, Clone, Default)]
+pub struct CommentSpan {
+    // The byte position in the string of the start of the span.
+    pub start: usize,
+    // The byte position in the string of the end of the span.
+    pub end: usize,
+}
+
+impl Ord for CommentSpan {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // If the starting position is the same encapsulatig span (i.e, wider one) should come
+        // first
+        match self.start.cmp(&other.start) {
+            Ordering::Equal => self.end.cmp(&other.end),
+            ord => ord,
+        }
+    }
+}
+
+impl PartialOrd for CommentSpan {
+    fn partial_cmp(&self, other: &CommentSpan) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub type CommentMap = BTreeMap<CommentSpan, Comment>;
+
+/// Get the CommentedTokenStream and collect the spans -> Comment mapping for the input source
+/// code.
+pub fn comment_map_from_src(input: Arc<str>) -> Result<CommentMap, FormatterError> {
+    let mut comment_map = BTreeMap::new();
+
+    // pass the input through lexer
+    let commented_token_stream = lex_commented(&input, 0, input.len(), None)?;
+    let tts = commented_token_stream.token_trees().iter();
+
+    for comment in tts {
+        collect_comments_from_token_stream(comment, &mut comment_map);
+    }
+    Ok(comment_map)
+}
+
+/// Collects `Comment`s from the token stream and insert it with its span to the `CommentMap`.
+/// Handles both the standalone and in-block comments.
+fn collect_comments_from_token_stream(
+    commented_token_tree: &CommentedTokenTree,
+    comment_map: &mut CommentMap,
+) {
+    match commented_token_tree {
+        CommentedTokenTree::Comment(comment) => {
+            let comment_span = CommentSpan {
+                start: comment.span.start(),
+                end: comment.span.end(),
+            };
+            comment_map.insert(comment_span, comment.clone());
+        }
+        CommentedTokenTree::Tree(CommentedTree::Group(group)) => {
+            for item in group.token_stream.token_trees().iter() {
+                collect_comments_from_token_stream(item, comment_map);
+            }
+        }
+        _ => {}
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{comment_map_from_src, CommentSpan};
+    use std::{ops::Bound::Included, sync::Arc};
+
+    #[test]
+    fn test_comment_span_ordering() {
+        let first_span = CommentSpan { start: 2, end: 4 };
+        let second_span = CommentSpan { start: 2, end: 6 };
```

Comment:
Oh, I just saw your response to my Ordering suggestion. It looks like this test is incorrect - I think the `start: 2, end: 6` span should occur *before* the `start: 2, end: 4` span, as I think we want the "visit" order where the enclosing spans precede the enclosed spans. Does that make sense?